modules/storages/app/services/storages/upload_file_service.rb

```diff
@@ -0,0 +1,173 @@
+
+    def call(file_path:, filename:, file_data:)
+      with_tagged_logger do
+        info "Starting file upload for #{filename} to #{file_path}"
+
+        unless nextcloud_storage?
+          @result.errors.add(:base, :unsupported_storage_type)
+          @result.success = false
+          return @result
+        end
+
+        ensure_validity_period_exists
+
+        folder = fetch_or_create_folder(file_path).value_or { return @result }
+        file = upload_file(folder, filename, file_data).value_or { return @result }
+        file_link = create_file_link(file).value_or { return @result }
+
+        @result.success = file_link
+        @result
+      end
+    end
+
+    private
+
+    def auth_strategy = Adapters::Registry["nextcloud.authentication.userless"].call
+
+    def nextcloud_storage?
+      @storage.short_provider_type == "nextcloud"
+    end
+
+    def error_with_code(error, code)
+      Adapters::Results::Error.new(payload: error.payload, source: self.class).with(code: code)
+    end
+
+    def determine_user(container)
+      if container.is_a?(WorkPackage) && container.author.present?
+        container.author
+      else
+        User.system
+      end
+    end
+
+    # TODO: do we need to validate validity_period here? For some reason FileLink creation fails otherwise.
+    def ensure_validity_period_exists
+      if @container.respond_to?(:start_date) && @container.respond_to?(:due_date)
+        @container.start_date ||= Time.zone.today
+        @container.due_date ||= Time.zone.today + 7.days
+      end
+
+      @container.save! if @container.changed?
+    end
+
+    def fetch_or_create_folder(file_path)
+      prefix = @project_storage.managed_project_folder_path
+      current_folder = nil
+
+      cumulative_paths(prefix, file_path).each do |folder_path|
+        current_folder = check_folder_exists?(folder_path).value_or do |error|
+          if error.code == :not_found
+            create_folder!(folder_path)
+          else
+            return Failure(error_with_code(error, error.code))
+          end
+        end
+      end
+
+      # TODO: file query returns collection, create folder returns monad result
+      parent_folder = current_folder.respond_to?(:success?) ? current_folder.value! : current_folder.parent
+      Success(parent_folder)
+    end
+
+    def cumulative_paths(prefix, path)
+      segments = path.split("/").reject(&:empty?)
+      return [prefix] if segments.empty?
+
+      segments.map.with_index { |_, i| "#{prefix}#{segments[0..i].join('/')}" }
+    end
+
+    def check_folder_exists?(path)
+      input_data = Adapters::Input::Files.build(folder: path).value_or do |error|
+        add_validation_error(error, options: { folder: path })
+      end
+
+      Adapters::Registry.resolve("#{@storage}.queries.files").call(auth_strategy:, storage: @storage, input_data:)
+    end
+
+    def create_folder!(path)
+      folder_path = File.dirname(path)
+      folder_name = path.sub("#{folder_path}/", "")
```

Comment:
:green_circle: I'd softly prefer `delete_prefix` as its intention is a bit clearer. Though the behaviour is the same, so it doesn't matter too much:

```suggestion
      folder_name = path.delete_prefix("#{folder_path}/")
```